src/osemgrep/cli_ci/Ci_subcommand.ml

```diff
@@ -322,15 +322,23 @@ let generate_meta_from_environment caps (baseline_ref : Digestif.SHA1.t option)
 let is_blocking (json : JSON.t) =
```

Comment:
oh nice, looking similar to the python side. Some issues here:

* in `partition_rules` we pass the rule metadata to `is_blocking`. this means we still need the param to `json JSON.t`. We either have to take into account that validation_state might not exist, or have a different is_blocking method when checking just rules (and not rule matches)

Also as is, doesn't compile, but editing your suggestion with the bullet point ^ in mind.

Reply:
Not as concise as yours but this ended up compiling. Please highlight any improvements that can be made!

```ocaml
let is_blocking (json : JSON.t) =
  match json with
  | JSON.Object xs ->
      (* Extract the validation state if present *)
      let validation_state = 
        match List.assoc_opt "validation_state" xs with
        | Some (JSON.String s) -> json_string_to_validation_state s
        | _ -> None
      in

      (* Determine if the validation state leads to blocking, based on actions *)
      let validation_state_should_block =
        match List.assoc_opt "dev.semgrep.validation_state.actions" xs with
        | Some (JSON.Object vs) -> (
            match validation_state with
            | Some `Confirmed_invalid | Some `Validation_error ->
                (match List.assoc_opt "invalid" vs, List.assoc_opt "error" vs with
                  | _, Some (JSON.String s) when String.equal s "block" -> true
                  | Some (JSON.String s), _ when String.equal s "block" -> true
                  | _ -> false)
            | Some `Confirmed_valid | Some `No_validator ->
                (match List.assoc_opt "valid" vs with
                  | Some (JSON.String s) when String.equal s "block" -> true
                  | _ -> false)
            | _ -> false
          )
        | _ -> false
      in

      (* Check for block actions outside the validation state actions *)
      let general_should_block =
        match List.assoc_opt "dev.semgrep.actions" xs with
        | Some (JSON.Array stuff) ->
            List.exists
              (function
                | JSON.String s -> String.equal s "block"
                | _ -> false)
              stuff
        | _ -> false
      in
      validation_state_should_block || general_should_block
  | _ -> f
```

Reply:
I think I fixed your original implementation:

```ocaml
let is_blocking (json: JSON.t) =
    let contains_blocking xs =
      List.exists
        (function
          | JSON.String s -> String.equal s "block"
          | _ -> false)
        xs
    in
    match json with
    | JSON.Object xs ->
        let validation_state_should_block =
          match
            ( List.assoc_opt "validation_state" xs,
              List.assoc_opt "dev.semgrep.validation_state.actions" xs )
          with
          | (Some (JSON.String validation_state), Some (JSON.Object actions)) ->
              (match
                 (validation_state, List.assoc_opt validation_state actions)
               with
               | (_, Some (JSON.String s)) -> String.equal s "block"
               | _ -> false)
          | _ -> false
        in
        let should_block =
          match List.assoc_opt "dev.semgrep.actions" xs with
          | Some (JSON.Array actions) -> contains_blocking actions
          | _ -> false
        in
        validation_state_should_block || should_block
    | _ -> false
```

Reply:
> in partition_rules we pass the rule metadata to is_blocking. this means we still need the param to json JSON.t. We either have to take into account that validation_state might not exist, or have a different is_blocking method when checking just rules (and not rule matches)

Yes, this would require updating the usage as well. I do not see how your revision would work, given I don't see why `validation_state` would be present in the metadata.

Reply:
Added a rule and match specific is blocking function in [45f852d](https://github.com/semgrep/semgrep/pull/9717/commits/45f852dee082b89352bf84861da504a46d464e5c)